### PR TITLE
Switch YAML parsing to Ruamel

### PR DESCRIPTION
**Pull request by rcoh**
- Change YAML parsing to Ruamel
- Wrap Ruamel output in `YamlTree` to preserve all the span information (Ruamel only preserves the start, not the end of spans in it's normal output)
- Coax the mypy gods into accepting this. Python is bad for ADTs...

Part of the #773 rebuild.

### CODE REVIEW

File: semgrep/semgrep/core_runner.py [hunk 1]
```diff
@@ -16,6 +16,8 @@
 from typing import Optional
 from typing import Tuple
 
+from ruamel.yaml import YAML
```
Comment (brendongo) [inline semgrep/semgrep/core_runner.py:19]: Does ruamel.yaml have a fast import time wrt yaml?

Comment (rcoh) [reply semgrep/semgrep/core_runner.py:19]: I think we're good: `semgrep --version  0.22s user 0.01s system 100% cpu 0.232 total`

File: semgrep/semgrep/rule_lang.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    class SpanPreservingRuamelConstructor(RoundTripConstructor):
+        def construct_object(self, node: Node, deep: bool = False) -> YamlTree:
+            r = super().construct_object(node, deep)
+            return YamlTree(r, Span.from_node(node, filename))
+
+    yaml = YAML()
+    yaml.Constructor = SpanPreservingRuamelConstructor
+    data = yaml.load(StringIO(contents))
+    if not isinstance(data, YamlTree):
+        raise Exception("Something has gone horribly wrong in the YAML parser")
```
Comment (ievans) [inline semgrep/semgrep/rule_lang.py]: Can we use the standard {PLEASE_REPORT_ISSUE...} here?

File: semgrep/semgrep/rule_lang.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            return {str(k.unroll()): v.unroll() for k, v in self.value.items()}
+        elif isinstance(self.value, YamlTree):
+            return self.value.unroll()
+        else:
+            return self.value
+
+
+def parse_yaml(contents: str) -> Dict[str, Any]:
+    yaml = YAML()
+    return yaml.load(StringIO(contents))  # type: ignore
```
Comment (ievans) [inline semgrep/semgrep/rule_lang.py:78]: should we explicit note in comment or function name whether or not this is safe?

**Review by ievans — APPROVED** [review]

File: semgrep/semgrep/rule_lang.py (hunk 3 above)
Comment (ievans) [reply semgrep/semgrep/rule_lang.py:78]: (seems actually pretty important given that we suggest parsing remote YAML files all the time)

**Pull request merged**